### PR TITLE
Add pathmap pipes module for path matching using Aho-Corasick #711

### PR DESCRIPTION
**Pull request by tdruez**

### CODE REVIEW

File: scanpipe/pipes/d2d.py [hunk 1]
```diff
@@ -264,46 +265,46 @@ def get_diff_ratio(to_resource, from_resource):
...
+    matches_iterator = pathmap.find_paths(to_resource.path, from_resources_index)
+
+    matches = list(matches_iterator)
+    if not matches:
+        return
 
-    for path_parts_index in range(1, path_parts_len):
-        current_parts = path_parts[path_parts_index:]
-        current_path = "/".join(current_parts)
+    matched_path_length, resource_ids = matches[0]
```
Comment (pombredanne) [inline scanpipe/pipes/d2d.py:277]: Are you sure we only ever get a single match? I am not sure that there is any guarantee for this.

File: scanpipe/pipes/d2d.py [hunk 2]
```diff
@@ -143,59 +144,96 @@ def checksum_map(project, checksum_field, logger=None):
...
+            from_resource_count,
             last_percent,
             increment_percent=10,
             start_time=start_time,
         )
-        _resource_java_to_class_map(to_resource, from_resources)
+        map_from_resource_java_to_class(from_resource, to_resources, to_java_fqn_index)
 
     # Flag not mapped .class in to/ codebase
+    to_resources = project_files.to_codebase().has_no_relation()
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:208]: This seems un-needed.

File: scanpipe/pipes/d2d.py [hunk 3]
```diff
@@ -143,59 +144,96 @@ def checksum_map(project, checksum_field, logger=None):
...
     to_resources_dot_class = to_resources.filter(name__endswith=".class")
+    to_resources_count = to_resources_dot_classes.count()
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:210]: This seems un-needed.

File: scanpipe/pipes/d2d.py [hunk 4]
```diff
@@ -29,12 +29,15 @@
...
+from scanpipe.pipes import jvm
+from scanpipe.pipes import pathmap
 from scanpipe.pipes import purldb
+from scanpipe.pipes import scancode
 
 FROM = "from/"
 TO = "to/"
 
-IGNORE_FILENAMES = ("packageinfo",)
+IGNORE_FILENAMES = ("package-info",)
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:40]: We do not want to ignore "packageinfo" as well, right?

File: scanpipe/pipes/d2d.py [hunk 5]
```diff
@@ -177,34 +187,133 @@ def java_to_class_map(project, logger=None):
...
+    """
+    Yield tuples of (resource id, fully-qualified Java class name) for indexable
+    classes from the "from/" side of the project codebase using the
+    "java_package" Resource.extra_data.
+    """
+    resource_values = from_resources_dot_java.values_list("id", "name", "extra_data")
+    return get_indexable_qualified_java_paths_from_values(resource_values)
+
+
+def find_java_packages(project, logger=None):
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:254]: This needs a dedicated unit test in test_d2d.

File: scanpipe/pipes/jvm.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    path = Path(path.strip("/"))
+    class_name = path.name
+    if "$" in class_name:  # inner class
+        class_name, _, _ = class_name.partition("$")
+    else:
+        class_name, _, _ = class_name.partition(".")  # plain .class
+    return str(path.parent / f"{class_name}.java")
+
+
+def get_fully_qualified_java_path(java_package, filename):
```
Comment (tdruez) [inline scanpipe/pipes/jvm.py:107]: A simple unit test form the docstring would be useful.

**Review by tdruez — COMMENTED** [review]
@pombredanne The performance improvements for `map_java_to_class` and `map_path` are massive with the new Aho-Corasick index, from hours to minutes.
See my comments for small improvements before the merge.

File: scanpipe/pipes/d2d.py [hunk 7]
```diff
@@ -29,14 +29,17 @@
...
 from scanpipe.pipes import purldb
+from scanpipe.pipes import scancode
 
 FROM = "from/"
 TO = "to/"
 
-IGNORE_FILENAMES = ("packageinfo",)
-IGNORE_EXTENSIONS = ()
-IGNORE_PATHS = ("gradleTest/",)
+IGNORED_FILENAMES = ("package-info",)
```
Comment (pombredanne) [inline scanpipe/pipes/d2d.py:40]:
```suggestion
IGNORED_FILENAMES = ("package-info.java", "package-info.class",)
```

File: scanpipe/pipes/d2d.py [hunk 8]
```diff
@@ -140,34 +143,41 @@ def checksum_map(project, checksum_field, logger=None):
...
+    match = pathmap.find_paths(path=normalized_java_path, index=from_classes_index)
+    if not match:
+        return
 
-    matches = from_resources.filter(path__endswith=qualified_java)
-    for match in matches:
+    for resource_id in match.resource_ids:
+        from_resource = from_resources.get(id=resource_id)
+        # compute the root of the packages on the source side
+        # TODO: consider storing this?
```
[the remark this answers is not in the document]
Comment (pombredanne) [reply scanpipe/pipes/d2d.py:163]: None indeed

Comment (pombredanne) [inline scanpipe/pipes/d2d.py:163]:
```suggestion
```

**Pull request merged**